allen-cell-animated/volume-viewer
# Feature/load volume subsets

Adds support for volumes which contain a subset of their full source data, and enables `OMEZarrLoader` to take advantage of these features. This enables several future optimizations, most immediately the ability to load single slices (resolves #110).

### Changes to `OmeZarrLoader`

- `OmeZarrLoader` is now *stateful*: it caches information on its data source (multiscale sizes, metadata, etc.) when `createVolume` is called and assumes that `loadVolumeData` will only be called on this source.
- `createVolume` now also treats the extent fields of `LoadSpec` (`minx`, `maxy`, etc.) differently to `loadVolumeData`.
  - Fields passed into `createVolume` define the size of the `Volume` object relative to the size of the data source, and are cached for use in `loadVolumeData`. This could be used, for example, to load single cell views directly out of full field zarr images.
  - Fields passed into `loadVolumeData` define the size of the subset relative to the volume, as in the screenshot below.
- As before, `createVolume` and `loadVolumeData` pick a scale level based on a fixed memory limitation. This means requesting a smaller subset may mean automatically loading a larger scale level. Extent fields are therefore floats in the range 0-1 rather than pixel values to remain agnostic of scale level.

### Changes to `Volume`

Volumes are still assumed to have a fixed total size (i.e. the size of their bounding box), but may contain a smaller chunk of actual volume data. This chunk will be packed densely into the texture (that is, without inserting gaps to fill the entire volume). This works like this:
- The `ImageInfo` interface has new fields: `vol_size_`(`x`/`y`/`z`), which represents the total size of the volume; and `offset_`(`x`/`y`/`z`), which represents the offset of the subset into the total volume area. For backwards compatibility with `JsonImageInfoLoader`, which loads `ImageInfo` objects and atlased textures directly from a remote source, these new fields are optional. When not specified, `vol_size` fields default to the values of `tile_width`, `tile_height`, and `tiles`, which represent the size of the data; and `offset` fields default to 0.
- While I was there, I added docstrings to all `ImageInfo` fields for clarity, and removed the derivable `atlas_width` and `atlas_height` properties.
- When `ImageInfo` is updated, `Volume` derives new properties, all `Vector3`s in the range 0-1: `contentSize` and `contentOffset`, and the function `getContentCenter` which calculates the center point of the subset.
- `VolumeRenderImpl`s use the above fields to show volumes with subsets properly.

### Steps to verify

I tested this by manually modifying the `LoadSpec` in the test app. The place to do this is in public/index.ts, in `loadVolume`, between `loader.createVolume` and `loader.loadVolumeData` (line 1160). Drop some lines like `loadSpec.maxy = 0.9;` in there and see what happens. It's possible I could modify the dat.gui or tweakpane UI to add some reload capability and make testing easier, but I didn't consider it worth working out how that would work at the moment.

Note that, because channels load one at a time, doing a load of a different extent into a volume briefly puts the volume into a state where the dimensions of some channels don't align with the dimensions of the volume. This isn't ideal, but I didn't find this to be too disruptive in my testing. There's plenty we could do down the line to mitigate this effect, but I considered it out of scope for now.

### Screenshots

![image](https://github.com/allen-cell-animated/volume-viewer/assets/53030214/d8e96f03-f621-449c-888e-6cf74e5a1859)
![image](https://github.com/allen-cell-animated/volume-viewer/assets/53030214/0cc7af57-34e2-4939-a29f-66cb9d0e62b3)
